File: typescript/demo/error-codes-demo.ts
/**
 * TypeScript SDK - Error Codes Demo
 * 
 * This demo shows how error codes and source locations are included in validation errors.
 * 
 * Run with: npx tsx demo/error-codes-demo.ts
 */

import { SchemaValidator, InstanceValidator, ErrorCodes } from '../dist/index.mjs';

try {
console.log('=== JSON Structure TypeScript SDK - Error Codes Demo ===\n');

// Test 1: Instance validation errors with source location tracking
console.log('--- Instance Validation Errors ---\n');

const schemaJson = `{
    "$schema": "https://json-structure.org/meta/core/v0/#",
    "$id": "https://example.com/test",
    "name": "TestSchema",
    "type": "object",
    "properties": {
        "name": { "type": "string" },
        "age": { "type": "int32" }
    },
    "required": ["name", "age"]
}`;

const instanceJson = `{
    "name": 123,
    "extra": "property"
}`;

console.log('Instance JSON:');
const lines = instanceJson.split('\n');
lines.forEach((line, i) => console.log(`  Line ${i + 1}: ${line}`));
console.log();

// Parse both schema and instance
const schema = JSON.parse(schemaJson);
const instance = JSON.parse(instanceJson);

const instanceValidator = new InstanceValidator({});
// Pass parsed objects, with optional source string for location tracking
const instanceResult = instanceValidator.validate(instance, schema, instanceJson);

console.log(`Valid: ${instanceResult.isValid}\n`);
console.log('Errors with error codes and source location:');
console.log('-'.repeat(80));

for (const error of instanceResult.errors) {
    console.log(`  Error: ${error.message}`);
    console.log(`         → Code:     ${error.code}`);
    console.log(`         → Path:     ${error.path}`);
    if (error.location) {
        console.log(`         → Location: Line ${error.location.line}, Column ${error.location.column}`);
        
        // Show the actual line from the source
        if (error.location.line > 0 && error.location.line <= lines.length) {
            const sourceLine = lines[error.location.line - 1];
            console.log(`         → Source:   ${sourceLine}`);
            console.log(`                     ${' '.repeat(error.location.column - 1)}^`);
        }
    }
    console.log();
}

// Test 2: Schema validation errors
console.log('\n--- Schema Validation Errors ---\n');

const badSchemaJson = `{
    "$schema": "https://json-structure.org/meta/core/v0/#",
    "$id": "https://example.com/bad",
    "name": "BadSchema",
    "type": "object",
    "properties": {
        "count": { "type": "integer" }
    },
    "required": ["missing_property"]
}`;

console.log('Bad Schema:');
const schemaLines = badSchemaJson.split('\n');
schemaLines.forEach((line, i) => console.log(`  Line ${i + 1}: ${line}`));
console.log();

// Parse the schema first, then pass to validate with source string
const badSchema = JSON.parse(badSchemaJson);
const schemaValidator = new SchemaValidator({});
const schemaResult = schemaValidator.validate(badSchema, badSchemaJson);

console.log(`Valid: ${schemaResult.isValid}\n`);
console.log('Errors with error codes:');
console.log('-'.repeat(80));

for (const error of schemaResult.errors) {
    console.log(`  Error: ${error.message}`);
    console.log(`         → Code:     ${error.code}`);
    console.log(`         → Path:     ${error.path}`);
    if (error.location) {
        console.log(`         → Location: Line ${error.location.line}, Column ${error.location.column}`);
    }
    console.log();
}

// Test 3: Show available error codes
console.log('\n--- Available Error Codes (sample) ---\n');
console.log('Schema Error Codes:');
console.log(`  SCHEMA_NULL = "${ErrorCodes.SCHEMA_NULL}"`);
console.log(`  SCHEMA_INVALID_TYPE = "${ErrorCodes.SCHEMA_INVALID_TYPE}"`);
console.log(`  SCHEMA_REF_NOT_FOUND = "${ErrorCodes.SCHEMA_REF_NOT_FOUND}"`);
console.log(`  SCHEMA_MIN_GREATER_THAN_MAX = "${ErrorCodes.SCHEMA_MIN_GREATER_THAN_MAX}"`);

console.log('\nInstance Error Codes:');
console.log(`  INSTANCE_STRING_MIN_LENGTH = "${ErrorCodes.INSTANCE_STRING_MIN_LENGTH}"`);
console.log(`  INSTANCE_REQUIRED_PROPERTY_MISSING = "${ErrorCodes.INSTANCE_REQUIRED_PROPERTY_MISSING}"`);
console.log(`  INSTANCE_TYPE_MISMATCH = "${ErrorCodes.INSTANCE_TYPE_MISMATCH}"`);
console.log(`  INSTANCE_STRING_PATTERN_MISMATCH = "${ErrorCodes.INSTANCE_STRING_PATTERN_MISMATCH}"`);

console.log('\n=== Demo Complete ===');
} catch (error) {
    console.error('Error:', error);
    process.exit(1);
}

File: go/demo/error_codes_demo.go
// Go SDK - Error Codes Demo
//
// This demo shows how error codes and source locations are included in validation errors.
package main

import (
	"fmt"
	"strings"

	jsonstructure "github.com/json-structure/sdk/go"
)

func main() {
	fmt.Println("=== JSON Structure Go SDK - Error Codes Demo ===")
	fmt.Println()

	// Test 1: Instance validation errors
	fmt.Println("--- Instance Validation Errors ---")
	fmt.Println()

	schemaJSON := `{
    "$schema": "https://json-structure.org/meta/core/v0/#",
    "type": "object",
    "properties": {
        "name": { "type": "string" },
        "age": { "type": "int32" }
    },
    "required": ["name", "age"]
}`

	instanceJSON := `{
    "name": 123,
    "extra": "property"
}`

	fmt.Println("Instance JSON:")
	lines := strings.Split(instanceJSON, "\n")
	for i, line := range lines {
		fmt.Printf("  Line %d: %s\n", i+1, line)
	}
	fmt.Println()

	// Use ValidateJSON to get source locations (not Validate with parsed objects)
	instanceValidator := jsonstructure.NewInstanceValidator(nil)
	result, err := instanceValidator.ValidateJSON([]byte(instanceJSON), []byte(schemaJSON))
	if err != nil {
		fmt.Printf("Parse error: %v\n", err)
		return
	}

	fmt.Printf("Valid: %v\n\n", result.IsValid)
	fmt.Println("Errors with error codes:")
	fmt.Println(strings.Repeat("-", 80))

	for _, err := range result.Errors {
		fmt.Printf("  Error: %s\n", err.Message)
		fmt.Printf("         → Code:     %s\n", err.Code)
		fmt.Printf("         → Path:     %s\n", err.Path)
		if err.Location.IsKnown() {
			fmt.Printf("         → Location: Line %d, Column %d\n", err.Location.Line, err.Location.Column)
		}
		fmt.Println()
	}

	// Test 2: Schema validation errors
	fmt.Println()
	fmt.Println("--- Schema Validation Errors ---")
	fmt.Println()

	badSchemaJSON := `{
    "$schema": "https://json-structure.org/meta/core/v0/#",
    "type": "object",
    "properties": {
        "count": { "type": "integer" }
    },
    "required": ["missing_property"]
}`

	fmt.Println("Bad Schema:")
	schemaLines := strings.Split(badSchemaJSON, "\n")
	for i, line := range schemaLines {
		fmt.Printf("  Line %d: %s\n", i+1, line)
	}
	fmt.Println()

	// Use ValidateJSON to get source locations
	schemaValidator := jsonstructure.NewSchemaValidator(nil)
	schemaResult, err := schemaValidator.ValidateJSON([]byte(badSchemaJSON))
	if err != nil {
		fmt.Printf("Parse error: %v\n", err)
		return
	}

	fmt.Printf("Valid: %v\n\n", schemaResult.IsValid)
	fmt.Println("Errors with error codes:")
	fmt.Println(strings.Repeat("-", 80))

	for _, err := range schemaResult.Errors {
		fmt.Printf("  Error: %s\n", err.Message)
		fmt.Printf("         → Code:     %s\n", err.Code)
		fmt.Printf("         → Path:     %s\n", err.Path)
		if err.Location.IsKnown() {
			fmt.Printf("         → Location: Line %d, Column %d\n", err.Location.Line, err.Location.Column)
		}
		fmt.Println()
	}

	// Test 3: Show available error codes
	fmt.Println()
	fmt.Println("--- Available Error Codes (sample) ---")
	fmt.Println()
	fmt.Println("Schema Error Codes:")
	fmt.Printf("  SchemaNull = \"%s\"\n", jsonstructure.SchemaNull)
	fmt.Printf("  SchemaInvalidType = \"%s\"\n", jsonstructure.SchemaInvalidType)
	fmt.Printf("  SchemaRefNotFound = \"%s\"\n", jsonstructure.SchemaRefNotFound)
	fmt.Printf("  SchemaMinGreaterThanMax = \"%s\"\n", jsonstructure.SchemaMinGreaterThanMax)

	fmt.Println()
	fmt.Println("Instance Error Codes:")
	fmt.Printf("  InstanceStringMinLength = \"%s\"\n", jsonstructure.InstanceStringMinLength)
	fmt.Printf("  InstanceRequiredPropertyMissing = \"%s\"\n", jsonstructure.InstanceRequiredPropertyMissing)
	fmt.Printf("  InstanceTypeMismatch = \"%s\"\n", jsonstructure.InstanceTypeMismatch)
	fmt.Printf("  InstanceStringPatternMismatch = \"%s\"\n", jsonstructure.InstanceStringPatternMismatch)

	fmt.Println()
	fmt.Println("=== Demo Complete ===")
}
